## Wrangle and Analyze Data

In this project a dataset from Twitter profile WeRateDogs will be wrangled (gather, assess and clean). All the efforts for this task are going to be reported in this notebook. 

### Name: Daniel Guarino

## Table of Contents

- [Introduction](#intro)
- [Basic Libraries](#libraries)
- [Part I - Gathering Data](#gathering)
- [Part II - Assessing Data](#assess)
- [Part III - Cleaning Data](#regression)


<a id='intro'></a>
### Introduction

Wrangling data is very important.....

<a id='libraries'></a>
### Import Basic Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

<a id='gathering'></a>
### Part I - Gathering Data

**Reading CSV file (manually downloaded from 'Resources' folder on Udacity - Wrangle and Analyze Data web page)**

In [2]:
df = pd.read_csv('twitter_archive/twitter-archive-enhanced.csv')

In [3]:
df.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

**Downloading a file programmatically**

In [4]:
import requests
import os

In [5]:
folder_name = 'twitter_archive'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [6]:
with open(os.path.join(folder_name, 
                       url.split('/')[-1]), mode = 'wb') as file:
    file.write(response.content)

In [7]:
os.listdir(folder_name)

['twitter-archive-enhanced.csv', 'image-predictions.tsv', 'tweet_json.txt']

In [8]:
df_image = pd.read_csv('twitter_archive/image-predictions.tsv', sep='\t')
df_image.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

**Importing and accessing Twitter API (tweep)**

In [9]:
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

In [11]:
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True)

In [12]:
tweet_ids = df.tweet_id.values
len(tweet_ids)

2356

In [ ]:
# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('twitter_archive/tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

# Cell output hidden to save lines

In [10]:
# Converting relevant JSON data to dataframe

tweets_list =[]

with open('twitter_archive/tweet_json.txt') as json_file:
    for line in json_file:
    
        tweets_dict = {}
        tweets_json = json.loads(line)
        
        try:
            tweets_dict['tweet_id'] = tweets_json['extended_entities']['media'][0]['id']
        except:
            tweets_dict['tweet_id'] = 'na'

        tweets_dict['retweet_count'] = tweets_json['retweet_count']
        tweets_dict['favorite_count'] = tweets_json['favorite_count']
        
        tweets_list.append(tweets_dict)

In [11]:
df_api = pd.DataFrame(tweets_list)
df_api.head()

tweet_id  retweet_count  favorite_count
0  892420639486877696           7947           37149
1  892177413194625024           5902           31957
2  891815175371796480           3902           24066
3  891689552724799489           8099           40478
4  891327551943041024           8785           38706

<a id='assess'></a>
### Part II - Assessing Data

### 1. Downloaded data by WeRateDogs page data frame info ###

In [12]:
df.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [14]:
df.dtypes

tweet_id                        int64
in_reply_to_status_id         float64
in_reply_to_user_id           float64
timestamp                      object
source                         object
text                           object
retweeted_status_id           float64
retweeted_status_user_id      float64
retweeted_status_timestamp     object
expanded_urls                  object
rating_numerator                int64
rating_denominator              int64
name                           object
doggo                          object
floofer                        object
pupper                         object
puppo                          object
dtype: object

In [15]:
# is there any duplicated ID?
sum(df['tweet_id'].duplicated())

0

In [17]:
# number of tweets where the rating numerator is smaller than the denominator
sum(df['rating_numerator'] < df['rating_denominator'])

442

In [18]:
df[df['rating_numerator'] < df['rating_denominator']]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
45    883482846933004288                    NaN                  NaN   
229   848212111729840128                    NaN                  NaN   
315   835152434251116546                    NaN                  NaN   
342   832088576586297345           8.320875e+17         3.058208e+07   
387   826598799820865537           8.265984e+17         4.196984e+09   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
45    2017-07-08 00:28:19 +0000   
229   2017-04-01 16:35:01 +0000   
315   2017-02-24 15:40:31 +0000   
342   2017-02-16 04:45:50 +0000   
387   2017-02-01 01:11:25 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
45    <a href="http://twitter.com/download/iphone" r...   
229   <a href="http://twitter.com/download/iphone" r...   
315   <a href="http://twitter.com/download/iphone" r...   
342   <a href="http://twitter.com/download/iphone" r...   
387   <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
45    This is Bella. She hopes her smile made you sm...                  NaN   
229   This is Jerry. He's doing a distinguished tong...                  NaN   
315   When you're so blinded by your systematic plag...                  NaN   
342            @docmisterio account started on 11/15/15                  NaN   
387   I was going to do 007/10, but the joke wasn't ...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
45                         NaN                        NaN   
229                        NaN                        NaN   
315                        NaN                        NaN   
342                        NaN                        NaN   
387                        NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
45    https://twitter.com/dog_rates/status/883482846...                 5   
229   https://twitter.com/dog_rates/status/848212111...                 6   
315   https://twitter.com/dog_rates/status/835152434...                 0   
342                

In [19]:
# example of numerator smaller than the denominator 
df['expanded_urls'][45]

'https://twitter.com/dog_rates/status/883482846933004288/photo/1,https://twitter.com/dog_rates/status/883482846933004288/photo/1'

In [20]:
# the numerator is 5 on the DataFrame, but it's 13.5 in the text
df['text'][45]

'This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948'

In [21]:
# is this tweet have more than one image?
df['tweet_id'][45]

883482846933004288

In [22]:
# no just one image
df_image[df_image['tweet_id'] == 883482846933004288]

tweet_id                                          jpg_url  \
2033  883482846933004288  https://pbs.twimg.com/media/DELC9dZXUAADqUk.jpg   

      img_num                p1   p1_conf  p1_dog                  p2  \
2033        1  golden_retriever  0.943082    True  Labrador_retriever   

       p2_conf  p2_dog      p3   p3_conf  p3_dog  
2033  0.032409    True  kuvasz  0.005501    True

In [23]:
# numerator equals to 0
df['expanded_urls'][315]

'https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1'

In [25]:
# number of denominators different than 10
sum(df['rating_denominator'] != 10)

23

### Quality Issues ###

**1- Completeness:**

- Missing data at `expanded_urls` column
- Missing data at `name` column. Not every dog has a name in this dataset
- We pratically don't have data for `in_reply_to_status_id` , `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp` columns. These 05 columns are useless for this dataset. 

**2- Validity:**

- Column `tweet_id` should be a string variable instead integer
- Column `timestamp` should be a datetime variable instead string

**3 - Accuracy:**

- The `rating_numerator` column values should be always larger or equal the `rating_denominator` column values (= 10). This is not true for 442 records
- The `rating_denominator` column values should be always 10, but this is not true for 23 records

**4- Consistency:**

- This dataset misses a column with the division `rating_numerator` / `rating_denominator`. This column is really important to further analysis


### Tidiness Issues ###

- The `timestamp` column should have separated columns for date and hour
- We should have an unique column for Dog Stages with values like doggo, puppo, pupper, flooter or Nan (not every dog has this classification).

### 2. Downloaded programmatically neural network udacity image dataframe info ###

In [26]:
df_image.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

In [27]:
df_image.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [28]:
# number of tweets with more than 1 image
sum(df_image['img_num'] > 1)

295

In [29]:
df_image.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [30]:
df_image.dtypes

tweet_id      int64
jpg_url      object
img_num       int64
p1           object
p1_conf     float64
p1_dog         bool
p2           object
p2_conf     float64
p2_dog         bool
p3           object
p3_conf     float64
p3_dog         bool
dtype: object

In [31]:
sum(df_image.duplicated())

0

In [32]:
# check how many breeds were predicted as a dog breed
sum(df_image['p1_dog'] == True)

1532

In [33]:
# check how many breeds were predicted as not a dog breed
sum(df_image['p1_dog'] == False)

543

In [34]:
# what kind of "breeds" we have when the algorithm didn't predicted a dog?
df_image[df_image['p1_dog'] == False].head(10)

tweet_id                                          jpg_url  \
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
8   666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
17  666104133288665088  https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18  666268910803644416  https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21  666293911632134144  https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22  666337882303524864  https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
25  666362758909284353  https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
29  666411507551481857  https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
33  666430724426358785  https://pbs.twimg.com/media/CT-jNYqW4AAPi2M.jpg   
43  666776908487630848  https://pbs.twimg.com/media/CUDeDoWUYAAD-EM.jpg   

    img_num                p1   p1_conf  p1_dog                  p2   p2_conf  \
6         1        box_turtle  0.933012   False          mud_turtle  0.045885   
8         1     shopping_cart  0.962465   False     shopping_basket  0.014594   
17        1               hen  0.965932   False                cock  0.033919   
18        1  desktop_computer  0.086502   False                desk  0.085547   
21        1  three-toed_sloth  0.914671   False               otter  0.015250   
22        1                ox  0.416669   False        Newfoundland  0.278407   
25        1        guinea_pig  0.996496   False               skunk  0.002402   
29        1              coho  0.404640   False          barracouta  0.271485   
33        1             llama  0.505184   False       Irish_terrier  0.104109   
43        1         seat_belt  0.375057   False  miniature_pinscher  0.167175   

    p2_dog                p3   p3_conf  p3_dog  
6    False          terrapin  0.017885   False  
8    False  golden_retriever  0.007959    True  
17   False         partridge  0.000052   False  
18   False          bookcase  0.079480   False  
21   False    great_grey_owl  0.013207   False  
22    True       groenendael  0.102643    True  
25   False           hamster  0.000461   False  
29   False               gar  0.189945   False  
33    True             dingo  0.062071   False  
43    True         Chihuahua  0.086951    True

In [35]:
sum(df_image['p1_conf'] >= 0.5)

1239

In [36]:
# test consistency of the breeds in df_image DataFrame
df_dog = df_image[df_image['p1_dog'] == True]


to print full data frame from https://stackoverflow.com/questions/19124601/pretty-print-an-entire-pandas-series-dataframe

In [37]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(df_dog['p1'].value_counts())

golden_retriever                  150
Labrador_retriever                100
Pembroke                           89
Chihuahua                          83
pug                                57
chow                               44
Samoyed                            43
toy_poodle                         39
Pomeranian                         38
cocker_spaniel                     30
malamute                           30
French_bulldog                     26
Chesapeake_Bay_retriever           23
miniature_pinscher                 23
Staffordshire_bullterrier          20
Siberian_husky                     20
German_shepherd                    20
Cardigan                           19
Eskimo_dog                         18
Maltese_dog                        18
beagle                             18
Shetland_sheepdog                  18
Lakeland_terrier                   17
Rottweiler                         17
Shih-Tzu                           17
kuvasz                             16
Italian_grey

In [38]:
# haven't found different names for the same breed! we don't have consistency problem with this dataset
# we have some breed names starting with capital letters and some with small letters. It's not a big deal nobody 
# is gonna be arrested because of this!

### Quality Issues ###

**1- Completeness:**

- Missing records: in this dataset we have 2075 entries, less than 2356, the number of entries in the downloaded by WeRateDogs dataset (checked above)
  
**2- Validity:**

- We have the same problem with the variable type of the column `tweet_id`. It should be a string instead an integer number 

**3 - Accuracy:**

- Not every predicted breed is a dog breed (e.g.: llama, guinea_pig). Some predictions are an object, not an animal (e.g.: seat_belt, shopping_cart, desktop_computer). This mistake happens in 543 predictions
- At `p1_conf` we have some poor predictions, (a breed has probability smaller than 50% - threshold choosen by me). The udacity neural network hasn't found a confident result for 836 records


### Tidiness Issues ###

- We should have an unique breed name in each row, in a column named `breed` as a variable. So we don't need the others columns (`p2_...` and `p3_...`)

### 3. Tweeter API downloaded Json file dataframe info ###

In [32]:
df_api.head()

tweet_id  retweet_count  favorite_count
0  892420639486877696           7947           37149
1  892177413194625024           5902           31957
2  891815175371796480           3902           24066
3  891689552724799489           8099           40478
4  891327551943041024           8785           38706

In [33]:
df_api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2333 entries, 0 to 2332
Data columns (total 3 columns):
tweet_id          2333 non-null object
retweet_count     2333 non-null int64
favorite_count    2333 non-null int64
dtypes: int64(2), object(1)
memory usage: 54.8+ KB


### Quality Issues ###

**1- Completeness:**

- Here the only problem is the mismatch in the number of entries. In this dataset we have 2333 entries against 2075 entries in the images data set and 2356 entries in the main dataset.

### Tidiness Issues ###

No tidiness issues in this dataset!

<a id='regression'></a>
### Part III - Cleaning Data

### Copy of the Data Frames ####

In [41]:
df_clean = df.copy()
df_image_clean = df_image.copy()
df_api_clean = df_api.copy()

### Quality Issues Cleaning ###

### Completeness ###

#### 1. Define ####

Remove `in_reply_to_status_id` , `in_reply_to_user_id`, `source`,`retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp` columns from df DataFrame.

#### Code ####

In [42]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [44]:
drop_columns_list = ['in_reply_to_status_id', 'in_reply_to_user_id', 'source','retweeted_status_id', 
                     'retweeted_status_user_id', 'retweeted_status_timestamp']

df_clean.drop(columns = drop_columns_list, axis= 1, inplace = True)

#### Test ####

In [45]:
# Should have 11 columns instead 17
df_clean.shape[1] == 11

True

### Validity ###

#### Define ####

Fix Validity problems to df and df_image DataFrames using .astype(str) pandas command for `tweet_id` column and pd.to_datetime for `timestamp` column 

#### Code ####

In [82]:
df_clean['tweet_id'] = df_clean['tweet_id'].astype(str)

In [83]:
df_image_clean['tweet_id'] = df_image_clean['tweet_id'].astype(str)

In [84]:
df_clean['timestamp'] = pd.to_datetime(df_clean['timestamp'])

#### Test ####

In [85]:
# ok
df_clean.dtypes

tweet_id                           object
timestamp             datetime64[ns, UTC]
source                             object
text                               object
expanded_urls                      object
rating_numerator                    int64
rating_denominator                  int64
name                               object
doggo                              object
floofer                            object
pupper                             object
puppo                              object
dtype: object

In [86]:
# ok!
df_image_clean.dtypes

tweet_id     object
jpg_url      object
img_num       int64
p1           object
p1_conf     float64
p1_dog         bool
p2           object
p2_conf     float64
p2_dog         bool
p3           object
p3_conf     float64
p3_dog         bool
dtype: object

### Accuracy ###

#### Define ####

df_clean: 
- Drop the 442 rows where the numerator is smaller than the denominator
- Drop the 23 rows where the denominator is different from 10
  

To drop these rows use df_clean.drop(row_index, inplace = True)

#### Code ####

In [87]:
df_clean.shape[0]

2356

In [88]:
# accuracy problems - 
wrong_numerator_index_number_array = df_clean[df_clean['rating_numerator'] < df_clean['rating_denominator']].index
wrong_numerator_index_number_array

Int64Index([  45,  229,  315,  342,  387,  462,  485,  599,  605,  730,
            ...
            2342, 2344, 2346, 2347, 2349, 2351, 2352, 2353, 2354, 2355],
           dtype='int64', length=442)

In [89]:
# drop rows
df_clean.drop(wrong_numerator_index_number_array, inplace = True)

In [90]:
wrong_denominator_index_number_array = df_clean[df_clean['rating_denominator'] != 10].index
wrong_denominator_index_number_array

Int64Index([ 313,  433,  516,  902, 1120, 1202, 1228, 1254, 1351, 1433, 1634,
            1635, 1663, 1779, 1843],
           dtype='int64')

In [91]:
# drop rows
df_clean.drop(wrong_denominator_index_number_array, inplace = True)

#### Test ####

In [92]:
sum(df_clean['rating_numerator'] < df_clean['rating_denominator']) == 0

True

In [93]:
sum(df_clean['rating_denominator'] != 10) == 0 

True

In [94]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1899 entries, 0 to 2350
Data columns (total 12 columns):
tweet_id              1899 non-null object
timestamp             1899 non-null datetime64[ns, UTC]
source                1899 non-null object
text                  1899 non-null object
expanded_urls         1853 non-null object
rating_numerator      1899 non-null int64
rating_denominator    1899 non-null int64
name                  1899 non-null object
doggo                 1899 non-null object
floofer               1899 non-null object
pupper                1899 non-null object
puppo                 1899 non-null object
dtypes: datetime64[ns, UTC](1), int64(2), object(9)
memory usage: 192.9+ KB


In [95]:
# entreies dropped from 2356 to 1899

#### Define ####

Drop rows where we don't have a dog breed and drop rows where the udacity neural networl algo didn't make a good prediction

#### Code ####

In [96]:
false_dog_breed_index = df_image_clean[df_image_clean['p1_dog'] == False].index
false_dog_breed_index

Int64Index([   6,    8,   17,   18,   21,   22,   25,   29,   33,   43,
            ...
            1992, 2012, 2013, 2021, 2022, 2026, 2046, 2052, 2071, 2074],
           dtype='int64', length=543)

In [97]:
df_image_clean.drop(false_dog_breed_index, inplace = True)

In [98]:
bad_dog_breed_prediction = df_image_clean[df_image_clean['p1_conf'] < 0.5].index
bad_dog_breed_prediction

Int64Index([   0,    3,    9,   12,   13,   14,   16,   19,   24,   26,
            ...
            2039, 2042, 2047, 2050, 2057, 2063, 2066, 2068, 2069, 2073],
           dtype='int64', length=559)

In [99]:
df_image_clean.drop(bad_dog_breed_prediction, inplace = True)

#### Test ####

In [100]:
sum(df_image_clean['p1_dog'] == False) == 0

True

In [101]:
sum(df_image_clean['p1_conf'] < 0.5) == 0

True

In [102]:
df_image_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 973 entries, 1 to 2072
Data columns (total 12 columns):
tweet_id    973 non-null object
jpg_url     973 non-null object
img_num     973 non-null int64
p1          973 non-null object
p1_conf     973 non-null float64
p1_dog      973 non-null bool
p2          973 non-null object
p2_conf     973 non-null float64
p2_dog      973 non-null bool
p3          973 non-null object
p3_conf     973 non-null float64
p3_dog      973 non-null bool
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 78.9+ KB


In [103]:
# entries numbers dropped from 2075 to 973

### Consistency ###

#### Define ####

df_clean:

- Create a new column `dog_rating` calculating `rating_numerator` / `rating_denominator` 

#### Code ####

In [104]:
# create new column
df_clean['dog_rating'] = df_clean['rating_numerator'] / df_clean['rating_denominator']

#### Test ####

In [105]:
# new column results
df_clean['dog_rating'].sample(25)

523      1.2
661      1.1
1209     1.0
693      1.2
1457     1.2
1151     1.2
306      1.3
743      1.1
954      1.1
734      1.1
264      1.2
38       1.2
1445     1.0
1098     1.0
2108     1.1
2074    42.0
955      1.3
2025     1.1
1196     1.2
686      1.1
157      1.3
364      1.2
676      1.3
2019     1.1
361      1.2
Name: dog_rating, dtype: float64

#### issue at 2074 row df_clean[dog_rating] is 42. Accuracy problem that must to be investigated in the Data Assement second iteration ####

### Tidiness ###

#### 1. Define ####

#### Code ####

In [120]:
df_clean.iloc[:,-6:-2].sample(20)

doggo floofer  pupper  puppo
232    None    None    None   None
840    None    None    None   None
753    None    None    None   None
94     None    None    None  puppo
871    None    None    None   None
1659   None    None    None   None
1595   None    None    None   None
299    None    None    None   None
312    None    None    None   None
679    None    None    None   None
1487   None    None  pupper   None
678    None    None    None   None
1859   None    None    None   None
1231   None    None    None   None
363   doggo    None    None   None
448   doggo    None    None   None
1506   None    None    None   None
1242   None    None    None   None
1523   None    None    None   None
116    None    None    None   None

In [123]:
# create a new column named dog_stage with blank data and check 
df_clean['dog_stage'] = np.nan
df_clean['dog_stage'].head(10)

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
5   NaN
6   NaN
7   NaN
8   NaN
9   NaN
Name: dog_stage, dtype: float64

In [118]:
df_clean.iloc[: , -6:-2]

doggo floofer pupper puppo
0     None    None   None  None
1     None    None   None  None
2     None    None   None  None
3     None    None   None  None
4     None    None   None  None
...    ...     ...    ...   ...
2339  None    None   None  None
2343  None    None   None  None
2345  None    None   None  None
2348  None    None   None  None
2350  None    None   None  None

[1899 rows x 4 columns]

In [127]:
# resolve this just use pd.melt 
for index in df_clean.iloc[: , -6:-2]:
    if index == 'doggo':
        if df_clean['doggo'] == 'doggo':
            df_clean['dog_stage'] == 'doggo'
        elif df_clean['dog_stage'] == 'None':
            df_clean['dog_stage'] == np.nan

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().